### DISCUSSION TITLE
Confused about the variable "always_taken"

### DISCUSSION THREAD

**Discussion by Mr-Cong** [ChampSim/ChampSim#346] [Q&A] [answered]
I‘m confused about the variable "always_taken"
1. The code in ooo_cpu.c
```C++
    if ((branch_prediction == 0) && (always_taken == 0)) {
      predicted_branch_target = 0;
    }
```
the logic is confusing.I think predicted_branch_target ==0 means misprediction.But if a condition branch is not always taken in BTB,and it is predicted as not taken in the branch predictor, the code will lead to a misprediction. I think it may be a bug.
A misprediction contains direction wrong,target wrong and both wrong.So in my opinion, the correct logic should use branch_prediction,always_taken to update the predicted_branch_target and then use it to compare with arch_instr.branch_target. 
2. The code in  basic_btb.c
```C++
    if (btb_entry == NULL) {
      // no prediction for this IP
      always_taken = true;
      return std::make_pair(0, always_taken);
    }
```
I guess the always_taken here is not important because here it seems to make no sense.But I have another guess. In fact I find the variable always_taken has not been well utilized .If we change the design to another case , in which we only use a branch to update the branch predictor when the always_taken is false, the situation in 2 can be explained.I'm not sure about it.

Regards

**Comment by alberto-ros** [comment, marked as answer]
Hi, the code you mention is from the master branch. A lot of discussion and some fixed were added to the develop branch regarding target prediction and the always_taken flag (e.g.., #318 #320 #322), so maybe you want to take a look at them. To clarify a bit your questions predicted_branch_target is set to 0 when the prediction is not taken (not when it is a misprediction).
